reconcile/test/test_utils_state.py

```diff
@@ -0,0 +1,65 @@
+import os
+from tempfile import NamedTemporaryFile
+
+import pytest
+import boto3
+from moto import mock_s3
+from reconcile.utils.state import State
+
+
+@pytest.fixture
+def accounts():
+    """Account name is needed to instantiate a State"""
+    return [{'name': 'some-account'}]
+
+
+@pytest.fixture
+def aws_credentials():
+    """Mocked AWS Credentials for moto."""
+    os.environ['AWS_ACCESS_KEY_ID'] = 'testing'
+    os.environ['AWS_SECRET_ACCESS_KEY'] = 'testing'
+    os.environ['AWS_SECURITY_TOKEN'] = 'testing'
+    os.environ['AWS_SESSION_TOKEN'] = 'testing'
+
+
+@mock_s3
+def test_ls(accounts, aws_credentials, mocker):
+    os.environ['APP_INTERFACE_STATE_BUCKET'] = 'some-bucket'
+    os.environ['APP_INTERFACE_STATE_BUCKET_ACCOUNT'] = 'some-account'
+    s3_client = boto3.client('s3', region_name='us-east-1')
+    s3_client.create_bucket(Bucket='some-bucket')
+
+    file_text = 'test'
+    with NamedTemporaryFile(delete=True, suffix='.txt') as tmp:
+        with open(tmp.name, 'w', encoding='UTF-8') as f:
+            f.write(file_text)
+
+        s3_client.upload_file(tmp.name, 'some-bucket',
+                              'state/integration-name/some-file-1')
+        """Creating some-file-2 to identify when two or more integrations have
+        similar names"""
+        s3_client.upload_file(tmp.name, 'some-bucket',
+                              'state/integration-name-2/some-file-2')
+
+    class MockAWSAPi:
+        def __init__(self, thread_pool_size, accounts,
+                     settings=None):
+            pass
+
+        @staticmethod
+        def get_session(account):
+            class Session:
+                @staticmethod
+                def client(type):
+                    return s3_client
+            return Session()
+
+    mocker.patch('reconcile.utils.state.AWSApi', MockAWSAPi)
```

Comment:
What is your thinking about taking this approach with creating `MockAWSAPi` versus using an [autospec mock](https://docs.python.org/3/library/unittest.mock.html#auto-speccing)?

Reply:
It was the first in the first examples for a google search for "pytest mock" 😅

Reply:
That's fair.  If you wanted to do this with Python mocks it would look something like this:

```python
mock_session = create_autospec(Session)

# This could also be mocker.patch.object(reconcile.utils.state, 'AWSApi', autospec=True)
mock_aws_api = mocker.patch('reconcile.utils.state.AWSApi', autospec=True)

mock_aws_api.return_value.get_session.return_value = mock_session
```

The nice thing about autospec is that it returns an object that matches the API of a real object, without needing to write and maintain additional classes.  If `Session` or `AWSApi` changed in this case, we'd need to update any stubs that have been created.  The nice thing about a stub or strict mock (see testing guide about testslide) is that any undefined behavior will fail, whereas MagicMocks will happily return for any method access.  This also makes the tests more prone to break where a new method is accessed.

For this use case, either way is probably fine because you really aren't using any other features that a mock provides (checking behavior), but for other more complex scenarios, a mock could be more appropriate.

Reply:
Thanks for the suggestions, I've pushed a new version.